lib/assert.js

```diff
@@ -285,15 +285,24 @@ function _deepEqual(actual, expected, strict, memos) {
   // Note: this accounts for both named and indexed properties on Arrays.
 
   // Use memos to handle cycles.
-  memos = memos || { actual: [], expected: [] };
-  const actualIndex = memos.actual.indexOf(actual);
-  if (actualIndex !== -1) {
-    if (actualIndex === memos.expected.indexOf(expected)) {
+  if (!memos) {
+    memos = {
+      actual: { map: new Map(), position: 0 },
+      expected: { map: new Map(), position: 0 }
+    };
+  }
+
+  const actualPosition = memos.actual.map.get(actual);
+  if (actualPosition !== undefined) {
+    if (actualPosition === memos.expected.map.get(expected)) {
```

Comment:
Changing from Array.prototype.indexOf (which uses [Strict Equality Comparison]) to Map.prototype.get (which uses [SameValueZero]) can have [subtle nuances](https://github.com/nodejs/node/pull/12142#issuecomment-290877915) that might result in incompatibilities, especially around NaN. Not sure if we need to care about it though.

[Strict Equality Comparison]: https://tc39.github.io/ecma262/#sec-strict-equality-comparison
[SameValueZero]: https://tc39.github.io/ecma262/#sec-samevaluezero